exonum-java-binding-core/rust/src/utils/jni_cache.rs

```diff
@@ -28,79 +43,91 @@ static mut TRANSACTION_EXECUTION_EXCEPTION: Option<GlobalRef> = None;
     JNI_VERSION_1_8
 }
 
-/// Caches all required classes and methods ids.
-pub fn cache_methods(env: &JNIEnv) {
+/// Initializes JNI cache implementing optimal synchronization
+pub fn init_cache(env: &JNIEnv) {
+    while CACHE_IS_LOCKED.compare_and_swap(false, true, Ordering::Acquire){}
+
     unsafe {
-        OBJECT_GET_CLASS =
-            get_method_id(&env, "java/lang/Object", "getClass", "()Ljava/lang/Class;");
-        CLASS_GET_NAME = get_method_id(&env, "java/lang/Class", "getName", "()Ljava/lang/String;");
-        THROWABLE_GET_MESSAGE = get_method_id(
-            &env,
-            "java/lang/Throwable",
-            "getMessage",
-            "()Ljava/lang/String;",
-        );
-        TRANSACTION_ADAPTER_EXECUTE = get_method_id(
-            &env,
-            "com/exonum/binding/service/adapters/UserTransactionAdapter",
-            "execute",
-            "(J)V",
-        );
-        TRANSACTION_ADAPTER_INFO = get_method_id(
-            &env,
-            "com/exonum/binding/service/adapters/UserTransactionAdapter",
-            "info",
-            "()Ljava/lang/String;",
-        );
-        TRANSACTION_ADAPTER_VERIFY = get_method_id(
-            &env,
-            "com/exonum/binding/service/adapters/UserTransactionAdapter",
-            "isValid",
-            "()Z",
-        );
-        SERVICE_ADAPTER_STATE_HASHES = get_method_id(
-            &env,
-            "com/exonum/binding/service/adapters/UserServiceAdapter",
-            "getStateHashes",
-            "(J)[[B",
-        );
-        SERVICE_ADAPTER_CONVERT_TRANSACTION = get_method_id(
-            &env,
-            "com/exonum/binding/service/adapters/UserServiceAdapter",
-            "convertTransaction",
-            "([B)Lcom/exonum/binding/service/adapters/UserTransactionAdapter;",
-        );
-        JAVA_LANG_ERROR = env
-            .new_global_ref(env.find_class("java/lang/Error").unwrap().into())
-            .ok();
-        TRANSACTION_EXECUTION_EXCEPTION = env
-            .new_global_ref(
-                env.find_class("com/exonum/binding/transaction/TransactionExecutionException")
-                    .unwrap()
-                    .into(),
-            ).ok();
-
-        assert!(
-            OBJECT_GET_CLASS.is_some()
-                && JAVA_LANG_ERROR.is_some()
-                && THROWABLE_GET_MESSAGE.is_some()
-                && TRANSACTION_ADAPTER_EXECUTE.is_some()
-                && TRANSACTION_ADAPTER_INFO.is_some()
-                && TRANSACTION_ADAPTER_VERIFY.is_some()
-                && SERVICE_ADAPTER_STATE_HASHES.is_some()
-                && SERVICE_ADAPTER_CONVERT_TRANSACTION.is_some()
-                && JAVA_LANG_ERROR.is_some()
-                && TRANSACTION_EXECUTION_EXCEPTION.is_some(),
-            "Error caching Java entities"
-        );
-
-        info!("Done caching references to Java classes and methods.");
+        if !CACHE_INITIALIZED {
+            cache_methods(env);
+            CACHE_INITIALIZED = true;
```

Comment:
Why is a spin lock used above? Can't we just perform an atomic write here (Release), and an atomic read before each access (Acquire) as we don't expect concurrent *initializations*?

Reply:
Even if we do expect, a lock would be required in initializer only; and it could be a simple lock, no need to DIY.

Reply:
Simplified it with parking_lot::Once that actually does exactly what we want (locks for initialization and atomic reads for accesing the state of initialization)